src/ui/actions/timeline.ts

```diff
@@ -467,65 +469,93 @@ export function setFocusRegion(focusRegion: FocusRegion | null): UIThunkAction {
-        dispatch(setTimelineToTime(startTime));
-      } else if (startTime === prevStartTime && endTime !== prevEndTime) {
-        dispatch(setTimelineToTime(endTime));
+    // Make sure our region is valid.
+    if (endTime < startTime) {
+      // If we need to adjust a dimension, it's the most intuitive to adjust the one that's being updated.
+      if (prevEndTime === endTime) {
+        startTime = endTime;
       } else {
-        // Else just make sure the preview time stays within the moving window.
-        const hoverTime = getHoverTime(state);
-        if (hoverTime !== null) {
-          if (hoverTime < startTime) {
-            dispatch(setTimelineToTime(startTime));
-          } else if (hoverTime > endTime) {
-            dispatch(setTimelineToTime(endTime));
-          }
-        } else {
-          dispatch(setTimelineToTime(currentTime));
-        }
+        endTime = startTime;
       }
+    }
 
-      dispatch(
-        setTrimRegion({
-          endTime,
-          startTime,
-        })
-      );
+    // Make sure the current time stays within the bounds of our selected region.
+    if (currentTime < startTime) {
+      dispatch(setTimelineState({ currentTime: startTime }));
+    } else if (currentTime > endTime) {
+      dispatch(setTimelineState({ currentTime: endTime }));
+    }
+
+    // Update the previous to match the handle that's being dragged.
+    if (startTime !== prevStartTime && endTime === prevEndTime) {
+      dispatch(setTimelineToTime(startTime));
+    } else if (startTime === prevStartTime && endTime !== prevEndTime) {
+      dispatch(setTimelineToTime(endTime));
     } else {
-      dispatch(setTrimRegion(null));
+      // Else just make sure the preview time stays within the moving window.
+      const hoverTime = getHoverTime(state);
+      if (hoverTime !== null) {
+        if (hoverTime < startTime) {
+          dispatch(setTimelineToTime(startTime));
+        } else if (hoverTime > endTime) {
+          dispatch(setTimelineToTime(endTime));
+        }
+      } else {
+        dispatch(setTimelineToTime(currentTime));
+      }
     }
+
+    // We now try and create a focus region with a time and a point, from just a
+    // time. In the future, it would be better if this were something like
+    // "setFocusRegionByTime" and we had another method which accepted
+    // TimeStampedPoints, because setting via context menu will often happen on
+    // resources where we already know the time *and* point. Also, we probably
+    // should *not* do this on every mouse movement in focus mode. This is
+    // actually foregoing most of the `getPointNearTime`, and instead relying on
+    // points that we mostly have already. I think often this will be OK, but
+    // probably we should also run "getPointNearTime", store that, and take the
+    // closest thing we can find of the points we have (assuming we don't have
+    // an exact match.)
```

Comment:
So I think the *ideal* case would be:

We are trying to translate a time to a point.

- We see if we have any exact matches (and if so, just use that, done).
- If we don't have any exact matches, go ask the backend for the nearest point, and then store the result (so next time we won't have to hit the backend)

However, that assumes that the points that the backend are returning are actually close enough to use (in some cases they are, in some they are not, the *average* amount that a backend response to `getPointNearTIme` right now is "off" is about a quarter of a second, but it has a long tail of higher errors like 10 seconds or more). We should improve the accuracy of that backend endpoint, and then we should centralize logic for the above, so we never have to wonder "which technique should I use here?"

However, we are not in that world yet, and I did not want to put a protocol call into something that I suspected might be getting hit on every mouse-move while setting the focus window, so I forewent the `getPointNearTime` here, for the reasons explained in the PR body. Let me know what you think here! I'm still trying to figure out the best way to do this!

Reply:
Yeah, that's fair. Thanks for elaborating.

I think in general, we should probably debounce some of this stuff (or use an approach like `useDeferredValue` or something). We definitely don't want to be firing off protocol messages on mouse-move.

I trust your intuition about whether this is the right approach. I was just uneasy because I wasn't sure how you'd gotten to that conclusion (or how I'd know to do the same thing if I were writing the code).